Optimizing ConditionAndOr queries

isuruperera opened the pull request:
converted queries with pattern ((A=1 AND B=2) OR (A=1 AND B=3)) to (A=1 AND (B=2 OR B=3))

isuruperera committed:
Optimizing ConditionAndOr queries according to the following pattern
convert ((A=1 AND B=2) OR (A=1 AND B=3)) to (A=1 AND (B=2 OR B=3))

grandinj commented on h2/src/main/org/h2/expression/ConditionAndOr.java:
```diff
@@ -300,4 +307,40 @@ public Expression getSubexpression(int index) {
+     * @return null or the third condition
+     */
+    private Expression optimizeConditionAndOr(ConditionAndOr left, ConditionAndOr right){
+        if(left.andOrType == AND && right.andOrType == AND){
```
I prefer the early out pattern, so please the method with:
    if (left.andOrType != AND || right.andOrType != AND) {
        return nullptr;
    }

isuruperera replied:
Resolved in the last commit

grandinj commented on h2/src/main/org/h2/expression/ConditionAndOr.java:
```diff
@@ -300,4 +307,40 @@ public Expression getSubexpression(int index) {
+        if(left.andOrType == AND && right.andOrType == AND){
+            Expression commonExpression;
+            Expression combinedExpression;
+            if(left.getSubexpression(0).getSQL().equals(right.getSubexpression(0).getSQL())){
```
left.getSubexpression(0).getSQL() and right.getSubexpression(0).getSQL() should be extracted into variables, so we only need to call them once

isuruperera replied:
Resolved in the last commit

grandinj commented on h2/src/main/org/h2/expression/ConditionAndOr.java:
```diff
@@ -300,4 +307,40 @@ public Expression getSubexpression(int index) {
+        }
+    }
+
 }
```
please add a test-case to org.h2.test.db.TestOptimizations

isuruperera replied:
I will add the test case after a few moments. Thank you for the suggestions.

isuruperera replied:
Added a test case

grandinj reviewed (changes requested):
thanks, just some small corrections needed!

isuruperera committed:
Optimizing ConditionAndOr queries according to the following pattern
convert ((A=1 AND B=2) OR (A=1 AND B=3)) to (A=1 AND (B=2 OR B=3))
-Suggested changes in:
https://github.com/h2database/h2database/pull/1571#discussion_r235925138
https://github.com/h2database/h2database/pull/1571#discussion_r235925223

isuruperera committed:
Bug fix in previous commit
Added test case

grandinj commented:
thank you, please can I get a license statement to the mailinglist/forum, something like mentioned in this link:
http://h2database.com/html/build.html#providing_patches

isuruperera commented:
> thank you, please can I get a license statement to the mailinglist/forum, something like mentioned in this link:
> http://h2database.com/html/build.html#providing_patches

I, Isuru Perera wrote the code included in this pull request, and it's mine, and I'm contributing it to H2 for distribution multiple-licensed under the MPL 2.0, and the EPL 1.0 (http://h2database.com/html/license.html).

grandinj commented:
please do a merge with master and resolve the conflicts

isuruperera committed:
Merge branch 'master' of https://github.com/h2database/h2database

isuruperera commented:
Merged with the master.

grandinj commented:
thank you!
